src/wmi.c

```diff
@@ -0,0 +1,324 @@
+    } else if (strcasecmp("InstancePrefix", c->key) == 0) {
+      cf_util_get_string(c, &instance_prefix);
+    } else if (strcasecmp("InstancesFrom", c->key) == 0) {
+      cf_util_get_string(c, &instances_from);
+    } else {
+      log_warn("ignoring unknown config key: \"%s\"", c->key);
+    }
+  }
+
+  if (metrics == NULL || statement == NULL ||
+      (instance_prefix == NULL && instances_from == NULL)) {
+    LIST_FREE(metrics, wmi_metric_free);
+    free(statement);
+    free(instance_prefix);
+    free(instances_from);
+    return NULL;
+  }
+
+  query = malloc(sizeof(wmi_query_t));
+  query->metrics = metrics;
+  query->statement = statement;
+  query->instance_prefix = instance_prefix;
+  query->instances_from = instances_from;
+  return query;
+}
+
+void wmi_query_free(wmi_query_t *q) {
+  if (!q)
+    return;
+
+  free(q->statement);
+  free(q->instance_prefix);
+  free(q->instances_from);
+  LIST_FREE(q->metrics, wmi_metric_free);
+  free(q);
+}
+
+void wmi_metric_free(wmi_metric_t *m) {
+  if (!m)
+    return;
+
+  free(m->type);
+  free(m->instance);
+  free(m->values_from);
+  free(m);
+}
+
+static void store(VARIANT *src, value_t *dst, int dst_type) {
+  switch (dst_type) {
+  case DS_TYPE_GAUGE:
+    dst->gauge = variant_get_double(src);
+    break;
+
+  case DS_TYPE_DERIVE:
+    dst->derive = variant_get_int64(src);
+    break;
+
+  case DS_TYPE_ABSOLUTE:
+    dst->absolute = variant_get_uint64(src);
+    break;
+
+  case DS_TYPE_COUNTER:
+    dst->counter = variant_get_uint64(src);
+    break;
+
+  default:
+    log_err("destination type '%d' is not supported", dst_type);
+    break;
+  }
+}
+
+/* Find position of `name` in `ds` */
+static int find_index_in_ds(const data_set_t *ds, const char *name) {
+  int i;
+  for (i = 0; i < ds->ds_num; i++)
+    if (strcmp(ds->ds[i].name, name) == 0)
+      return i;
+  return -1;
+}
+
+static int wmi_exec_query(wmi_query_t *q) {
+  wmi_result_list_t *results;
+  results = wmi_query(wmi, q->statement);
+
+  if (results->count == 0) {
+    log_warn("no results for query %s.", q->statement);
+    wmi_result_list_release(results);
+    return 0;
+  }
+
+  wmi_result_t *result;
+  while ((result = wmi_get_next_result(results))) {
+    value_list_t vl = VALUE_LIST_INIT;
+
+    sstrncpy(vl.host, hostname_g, sizeof(vl.host));
+    sstrncpy(vl.plugin, "wmi", sizeof(vl.plugin));
+
+    LIST_TYPE(wmi_metric_t) * mn;
+    for (mn = q->metrics; mn != NULL; mn = LIST_NEXT(mn)) {
+      VARIANT value_v;
+      VARIANT plugin_instance_v;
+      value_t value_vt;
+      char *plugin_instance_s = NULL;
+      const data_set_t *ds = NULL;
+      wmi_metric_t *m = LIST_NODE(mn);
+
+      ds = plugin_get_ds(m->type);
+      int index_in_ds;
+      if (wmi_result_get_value(result, m->values_from, &value_v) != 0) {
+        VariantClear(&value_v);
+        log_err("failed to read field '%s'", m->values_from);
+        continue;
+      }
+
+      index_in_ds = find_index_in_ds(ds, "value");
```

Comment:
This feels weird. While (now) common practice, there's no guarantee about the name of DSs. I think what you actually want here is to only support types with a single DS. I think there's prior art for that in other plugins. Also, this is something you could check at configure time.